hazelcast/src/main/java/com/hazelcast/core/FlakeIdGenerator.java

```diff
@@ -0,0 +1,155 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.core;
+
+import com.hazelcast.concurrent.flakeidgen.FlakeIdGeneratorDataSerializerHook;
+import com.hazelcast.internal.cluster.ClusterService;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Creates cluster-wide unique ID generator. Generated IDs are {@code long} primitive values
+ * and are k-ordered (roughly ordered). IDs are in the range from {@code Long.MIN_VALUE} to
+ * {@code Long.MAX_VALUE}.
+ * <p>
+ * The IDs contain timestamp component and a node ID, which is assigned when the member
+ * joins the cluster. This allows the IDs to be ordered and unique without any coordination between
+ * members, which makes the generator safe even in split-brain scenario (for caveats,
+ * {@link ClusterService#getMemberListJoinVersion() see here}).
+ * <p>
+ * Timestamp component is in milliseconds since 1.1.2017, 0:00 GMT, and has 42 bits. This caps
+ * the useful lifespan of the generator to little less than 140 years. The sequence component is 6 bits.
+ * If more than 64 IDs are requested in single millisecond, IDs will gracefully overflow to next
+ * millisecond and uniqueness is guaranteed in this case.
+ * <p>
+ * Node ID component of the ID has 16 bits. Members with member list join version higher than
+ * 2^16 won't be able to generate IDs. The clients will be able to generate IDs as long as there is at
+ * least one member with join version smaller than 2^16. The remedy is to restart the cluster:
+ * nodeId will be assigned from zero again. Uniqueness after the restart will be preserved thanks to
+ * the timestamp component.
+ *
+ * <h4>Performance</h4>
+ * Operation on member is always local. On client, the {@link #newId()} method goes to a random
+ * member and gets a batch of IDs, which will then be returned locally for limited time. The pre-fetch
+ * size and the validity time are configured per-client.
+ * The {@link #newIdBatch(int)} always goes to a member.
+ */
+public interface FlakeIdGenerator extends DistributedObject {
+
+    IdBatch newIdBatch(int batchSize);
+
+    /**
+     * Generates and returns a cluster-wide unique ID.
+     * <p>
+     * The call is always local on member. A batch of IDs is pre-fetched on the client and then used for
+     * preconfigured time locally. If you need multiple IDs at once, use {@link #newIdBatch(int)}.
+     * <p>
+     * <b>Note:</b> Values returned from this method can be not strictly ordered.
+     *
+     * @return new cluster-wide unique ID
+     */
+    long newId();
+
+    class IdBatch implements Iterable<Long>, IdentifiedDataSerializable {
```

Comment:
Yes, it's sent from member to client.

Reply:
@viliam-durina It should not send from member to client as IdentifiedDataSerializable. 
I think we already converted it to protocol. Removing `IdentifiedDataSerializable` should be OK right now.

Reply:
The client tests pass even if this class is not `IdentifiedDataSerializable` because you are acutally [encoding separate numbers as the response](https://github.com/hazelcast/hazelcast/blob/dcaeba65d51deb8a3c21629c38d7466ec158b8f3/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/flakeidgenerator/NewIdBatchMessageTask.java#L47) and [constructing the batch on the client side](https://github.com/hazelcast/hazelcast/blob/611c7505c8bcb310d00b987d4e3702610a298d1e/hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientFlakeIdGeneratorProxy.java#L63).

Reply:
I see now thank's to sancar's comments, I used `Data` during development.